controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java

```diff
@@ -462,16 +462,32 @@ public void listReaderGroups(final String scopeName, final SecurityContext secur
     public void listScopes(final SecurityContext securityContext, final AsyncResponse asyncResponse) {
         long traceId = LoggerHelpers.traceEnter(log, "listScopes");
 
+        try {
+            authenticate("/", READ);
+        } catch (AuthException e) {
+            log.warn("Get scopes failed due to authentication failure.");
+            asyncResponse.resume(Response.status(Status.fromStatusCode(e.getResponseCode())).build());
+            LoggerHelpers.traceLeave(log, "listScopes", traceId);
+            return;
+        }
+
         controllerService.listScopes()
-                .thenApply(scopesList -> {
-                    ScopesList scopes = new ScopesList();
-                    scopesList.forEach(scope -> scopes.addScopesItem(new ScopeProperty().scopeName(scope)));
-                    return Response.status(Status.OK).entity(scopes).build();
-                }).exceptionally(exception -> {
-                        log.warn("listScopes failed with exception: " + exception);
-                        return Response.status(Status.INTERNAL_SERVER_ERROR).build();
-                }).thenApply(asyncResponse::resume)
-                .thenAccept(x ->  LoggerHelpers.traceLeave(log, "listScopes", traceId));
+                         .thenApply(scopesList -> {
+                             ScopesList scopes = new ScopesList();
+                             scopesList.forEach(scope -> {
+                                 try {
+                                     authenticate(scope, READ);
```

Comment:
Could you please decouple `authenticate` and `authorize` so that `authenticate` is called once?   I thought that was the main point of the interface change.

Reply:
Fixed. My interpretation of the discussions was that we were planning to optimize this later. Updated with decoupled implementation.